// .ignored/@multiversx/sdk-core/out/transfers/transfersControllers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransfersController = void 0;
const core_1 = require("../core");
const transferTransactionsFactory_1 = require("./transferTransactionsFactory");
class TransfersController extends core_1.BaseController {
    constructor(options) {
        super({ gasLimitEstimator: options.gasLimitEstimator });
        this.factory = new transferTransactionsFactory_1.TransferTransactionsFactory({
            config: new core_1.TransactionsFactoryConfig(options),
        });
    }
    async createTransactionForNativeTokenTransfer(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForNativeTokenTransfer(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForEsdtTokenTransfer(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForESDTTokenTransfer(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForTransfer(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForTransfer(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
}
exports.TransfersController = TransfersController;
