// .ignored/@multiversx/sdk-core/out/core/baseController.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseController = void 0;
const address_1 = require("./address");
const constants_1 = require("./constants");
const transactionComputer_1 = require("./transactionComputer");
class BaseController {
    constructor(options) {
        this.gasLimitEstimator = options?.gasLimitEstimator;
    }
    addExtraGasLimitIfRequired(transaction) {
        if (transaction.guardian && !transaction.guardian.isEmpty()) {
            transaction.gasLimit += BigInt(constants_1.EXTRA_GAS_LIMIT_FOR_GUARDED_TRANSACTIONS);
        }
        if (transaction.relayer && !transaction.relayer.isEmpty()) {
            transaction.gasLimit += BigInt(constants_1.EXTRA_GAS_LIMIT_FOR_RELAYED_TRANSACTIONS);
        }
    }
    async setupAndSignTransaction(transaction, options, nonce, sender) {
        transaction.guardian = options.guardian ?? address_1.Address.empty();
        transaction.relayer = options.relayer ?? address_1.Address.empty();
        transaction.nonce = nonce;
        this.setVersionAndOptionsForGuardian(transaction);
        await this.setTransactionGasOptions(transaction, options);
        transaction.signature = await sender.signTransaction(transaction);
    }
    setVersionAndOptionsForGuardian(transaction) {
        if (transaction.guardian && !transaction.guardian.isEmpty()) {
            const txComputer = new transactionComputer_1.TransactionComputer();
            txComputer.applyGuardian(transaction, transaction.guardian);
        }
    }
    async setTransactionGasOptions(transaction, options) {
        if (options.gasPrice) {
            transaction.gasPrice = options.gasPrice;
        }
        if (options.gasLimit) {
            transaction.gasLimit = options.gasLimit;
            return;
        }
        this.addExtraGasLimitIfRequired(transaction);
        if (this.gasLimitEstimator) {
            transaction.gasLimit = await this.gasLimitEstimator.estimateGasLimit({ transaction });
        }
    }
}
exports.BaseController = BaseController;
