// .ignored/@multiversx/sdk-core/out/core/transactionEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionEvent = void 0;
const address_1 = require("./address");
class TransactionEvent {
    constructor(init) {
        this.raw = {};
        this.address = address_1.Address.empty();
        this.identifier = "";
        this.topics = [];
        this.data = new Uint8Array();
        this.additionalData = [];
        Object.assign(this, init);
    }
    static fromHttpResponse(responsePart) {
        let result = new TransactionEvent();
        result.address = new address_1.Address(responsePart.address);
        result.identifier = responsePart.identifier || "";
        result.topics = (responsePart.topics || []).map((topic) => Buffer.from(topic, "base64"));
        result.data = Buffer.from(responsePart.data ?? "", "base64");
        result.additionalData = (responsePart.additionalData || []).map((data) => Buffer.from(data, "base64"));
        result.raw = responsePart;
        return result;
    }
}
exports.TransactionEvent = TransactionEvent;
