// .ignored/@multiversx/sdk-core/out/accountManagement/accountController.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountController = void 0;
const core_1 = require("../core");
const transactionsFactoryConfig_1 = require("../core/transactionsFactoryConfig");
const accountTransactionsFactory_1 = require("./accountTransactionsFactory");
class AccountController extends core_1.BaseController {
    constructor(options) {
        super({ gasLimitEstimator: options.gasLimitEstimator });
        this.factory = new accountTransactionsFactory_1.AccountTransactionsFactory({
            config: new transactionsFactoryConfig_1.TransactionsFactoryConfig(options),
        });
    }
    async createTransactionForSavingKeyValue(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForSavingKeyValue(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForSettingGuardian(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForSettingGuardian(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForGuardingAccount(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForGuardingAccount(sender.address);
        transaction.relayer = options.relayer ?? core_1.Address.empty();
        transaction.nonce = nonce;
        await this.setTransactionGasOptions(transaction, options);
        transaction.signature = await sender.signTransaction(transaction);
        return transaction;
    }
    async createTransactionForUnguardingAccount(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForUnguardingAccount(sender.address, {
            guardian: options.guardian,
        });
        transaction.relayer = options.relayer ?? core_1.Address.empty();
        transaction.nonce = nonce;
        this.setVersionAndOptionsForGuardian(transaction);
        await this.setTransactionGasOptions(transaction, options);
        transaction.signature = await sender.signTransaction(transaction);
        return transaction;
    }
}
exports.AccountController = AccountController;
