// @multiversx/sdk-core@15.3.2/out/delegation/delegationController.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegationController = void 0;
const core_1 = require("../core");
const delegationTransactionsFactory_1 = require("./delegationTransactionsFactory");
const delegationTransactionsOutcomeParser_1 = require("./delegationTransactionsOutcomeParser");
class DelegationController extends core_1.BaseController {
    constructor(options) {
        super({ gasLimitEstimator: options.gasLimitEstimator });
        this.transactionAwaiter = new core_1.TransactionWatcher(options.networkProvider);
        this.factory = new delegationTransactionsFactory_1.DelegationTransactionsFactory({
            config: new core_1.TransactionsFactoryConfig({ chainID: options.chainID }),
        });
        this.parser = new delegationTransactionsOutcomeParser_1.DelegationTransactionsOutcomeParser();
    }
    async createTransactionForNewDelegationContract(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForNewDelegationContract(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async awaitCompletedCreateNewDelegationContract(txHash) {
        const transaction = await this.transactionAwaiter.awaitCompleted(txHash);
        return this.parseCreateNewDelegationContract(transaction);
    }
    parseCreateNewDelegationContract(transactionOnNetwork) {
        return this.parser.parseCreateNewDelegationContract(transactionOnNetwork);
    }
    async createTransactionForAddingNodes(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForAddingNodes(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForRemovingNodes(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForRemovingNodes(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForStakingNodes(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForStakingNodes(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForUnbondingNodes(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForUnbondingNodes(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForUnstakingNodes(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForUnstakingNodes(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForUnjailingNodes(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForUnjailingNodes(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForChangingServiceFee(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForChangingServiceFee(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForModifyingDelegationCap(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForModifyingDelegationCap(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForSettingAutomaticActivation(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForSettingAutomaticActivation(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForUnsettingAutomaticActivation(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForUnsettingAutomaticActivation(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForSettingCapCheckOnRedelegateRewards(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForSettingCapCheckOnRedelegateRewards(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForUnsettingCapCheckOnRedelegateRewards(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForUnsettingCapCheckOnRedelegateRewards(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForSettingMetadata(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForSettingMetadata(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForDelegating(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForDelegating(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForClaimingRewards(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForClaimingRewards(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForRedelegatingRewards(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForRedelegatingRewards(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForUndelegating(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForUndelegating(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
    async createTransactionForWithdrawing(sender, nonce, options) {
        const transaction = await this.factory.createTransactionForWithdrawing(sender.address, options);
        await this.setupAndSignTransaction(transaction, options, nonce, sender);
        return transaction;
    }
}
exports.DelegationController = DelegationController;
