// @multiversx/sdk-core@15.3.2/out/multisig/multisigTransactionsOutcomeParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultisigTransactionsOutcomeParser = void 0;
const transactionsOutcomeParsers_1 = require("../transactionsOutcomeParsers");
/**
 * Parses the outcome of multisig contract operations
 */
class MultisigTransactionsOutcomeParser {
    constructor(options) {
        this.abi = options?.abi;
        this.parser = new transactionsOutcomeParsers_1.SmartContractTransactionsOutcomeParser({ abi: this.abi });
    }
    /**
     * Parses the outcome of creating a new multisig contract
     * @param transactionOnNetwork The completed transaction
     * @returns An array of objects containing the new contract addresses
     */
    parseDeploy(transactionOnNetwork) {
        return this.parser.parseDeploy({ transactionOnNetwork });
    }
    /**
     * Parses the outcome of a multisig action proposal
     * @param transactionOnNetwork The completed transaction
     * @returns The action ID that was created
     */
    parseProposeAction(transactionOnNetwork) {
        const result = this.parser.parseExecute({ transactionOnNetwork });
        return result.values[0];
    }
    /**
     * Parses the outcome of a multisig action proposal
     * @param transactionOnNetwork The completed transaction
     * @returns In case of scDeploy returns address else undefined
     */
    parsePerformAction(transactionOnNetwork) {
        const result = this.parser.parseExecute({ transactionOnNetwork });
        return result.values[0];
    }
}
exports.MultisigTransactionsOutcomeParser = MultisigTransactionsOutcomeParser;
