// @multiversx/sdk-core@15.3.2/out/networkProviders/networkStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkStatus = void 0;
/**
 * An object holding network status configuration parameters.
 */
class NetworkStatus {
    constructor() {
        this.raw = {};
        this.currentRound = 0n;
        this.currentEpoch = 0;
        this.highestFinalNonce = 0n;
        this.blockNonce = 0n;
        this.blockTimestamp = 0;
    }
    /**
     * Constructs a configuration object from a HTTP response (as returned by the provider).
     */
    static fromHttpResponse(payload) {
        let networkStatus = new NetworkStatus();
        networkStatus.raw = payload;
        networkStatus.currentRound = BigInt(payload["erd_current_round"]);
        networkStatus.currentEpoch = Number(payload["erd_epoch_number"]);
        networkStatus.highestFinalNonce = BigInt(payload["erd_highest_final_nonce"]);
        networkStatus.blockNonce = BigInt(payload["erd_nonce"]);
        networkStatus.blockTimestamp = Number(payload["erd_block_timestamp"]);
        return networkStatus;
    }
}
exports.NetworkStatus = NetworkStatus;
