// @multiversx/sdk-core@15.3.2/out/networkProviders/resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionCostResponse = exports.AwaitingOptions = void 0;
const transactionStatus_1 = require("../core/transactionStatus");
const constants_1 = require("./constants");
class AwaitingOptions {
    constructor() {
        this.pollingIntervalInMilliseconds = constants_1.DEFAULT_ACCOUNT_AWAITING_POLLING_TIMEOUT_IN_MILLISECONDS;
        this.timeoutInMilliseconds = constants_1.DEFAULT_ACCOUNT_AWAITING_TIMEOUT_IN_MILLISECONDS;
        this.patienceInMilliseconds = constants_1.DEFAULT_ACCOUNT_AWAITING_PATIENCE_IN_MILLISECONDS;
    }
}
exports.AwaitingOptions = AwaitingOptions;
class TransactionCostResponse {
    constructor() {
        this.raw = {};
        this.gasLimit = 0;
        this.status = transactionStatus_1.TransactionStatus.createUnknown();
    }
    static fromHttpResponse(payload) {
        const result = new TransactionCostResponse();
        result.raw = payload;
        result.gasLimit = payload["txGasUnits"] ?? 0;
        result.status = new transactionStatus_1.TransactionStatus("");
        return result;
    }
}
exports.TransactionCostResponse = TransactionCostResponse;
