// .ignored/@multiversx/sdk-core/out/entrypoints/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalnetEntrypointConfig = exports.MainnetEntrypointConfig = exports.DevnetEntrypointConfig = exports.TestnetEntrypointConfig = void 0;
class TestnetEntrypointConfig {
    constructor({ networkProviderUrl = "https://testnet-api.multiversx.com", networkProviderKind = "api", chainId = "T", } = {}) {
        this.networkProviderUrl = networkProviderUrl;
        this.networkProviderKind = networkProviderKind;
        this.chainId = chainId;
    }
}
exports.TestnetEntrypointConfig = TestnetEntrypointConfig;
class DevnetEntrypointConfig {
    constructor({ networkProviderUrl = "https://devnet-api.multiversx.com", networkProviderKind = "api", chainId = "D", } = {}) {
        this.networkProviderUrl = networkProviderUrl;
        this.networkProviderKind = networkProviderKind;
        this.chainId = chainId;
    }
}
exports.DevnetEntrypointConfig = DevnetEntrypointConfig;
class MainnetEntrypointConfig {
    constructor({ networkProviderUrl = "https://api.multiversx.com", networkProviderKind = "api", chainId = "1", } = {}) {
        this.networkProviderUrl = networkProviderUrl;
        this.networkProviderKind = networkProviderKind;
        this.chainId = chainId;
    }
}
exports.MainnetEntrypointConfig = MainnetEntrypointConfig;
class LocalnetEntrypointConfig {
    constructor({ networkProviderUrl = "http://localhost:7950", networkProviderKind = "proxy", chainId = "localnet", } = {}) {
        this.networkProviderUrl = networkProviderUrl;
        this.networkProviderKind = networkProviderKind;
        this.chainId = chainId;
    }
}
exports.LocalnetEntrypointConfig = LocalnetEntrypointConfig;
