// .ignored/@multiversx/sdk-core/out/wallet/validatorSigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorSigner = void 0;
const errors_1 = require("../core/errors");
const validatorKeys_1 = require("./validatorKeys");
const validatorPem_1 = require("./validatorPem");
/**
 * Validator signer (BLS signer)
 */
class ValidatorSigner {
    constructor(secretKey) {
        this.secretKey = secretKey;
    }
    /**
     * * @deprecated This method will be deprecated! Use the sign method directly.
     * Signs a message.
     */
    static async signUsingPem(pemText, pemIndex = 0, signable) {
        await validatorKeys_1.BLS.initIfNecessary();
        try {
            let secretKey = validatorKeys_1.ValidatorSecretKey.fromPem(pemText, pemIndex);
            return secretKey.sign(signable);
        }
        catch (err) {
            throw new errors_1.ErrSignerCannotSign(err);
        }
    }
    static async fromPemFile(path, index = 0) {
        const secretKey = (await validatorPem_1.ValidatorPEM.fromFile(path, index)).secretKey;
        return new ValidatorSigner(secretKey);
    }
    sign(data) {
        try {
            return this.trySign(data);
        }
        catch (err) {
            throw new errors_1.ErrSignerCannotSign(err);
        }
    }
    trySign(data) {
        return this.secretKey.sign(data);
    }
    getPubkey() {
        return this.secretKey.generatePublicKey();
    }
}
exports.ValidatorSigner = ValidatorSigner;
