// .ignored/@multiversx/sdk-core/out/wallet/validatorPem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorPEM = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const pemEntry_1 = require("../wallet/pemEntry");
const validatorKeys_1 = require("../wallet/validatorKeys");
class ValidatorPEM {
    constructor(label, secretKey) {
        this.label = label;
        this.secretKey = secretKey;
    }
    static async fromFile(path, index = 0) {
        return (await this.fromFileAll(path))[index];
    }
    static async fromFileAll(path) {
        const absPath = path_1.resolve(path);
        const text = fs_1.readFileSync(absPath, "utf-8");
        return await this.fromTextAll(text);
    }
    static async fromText(text, index = 0) {
        return (await ValidatorPEM.fromTextAll(text))[index];
    }
    static async fromTextAll(text) {
        const entries = pemEntry_1.PemEntry.fromTextAll(text);
        const resultItems = [];
        await validatorKeys_1.BLS.initIfNecessary();
        for (const entry of entries) {
            const secretKey = new validatorKeys_1.ValidatorSecretKey(entry.message);
            const item = new ValidatorPEM(entry.label, secretKey);
            resultItems.push(item);
        }
        return resultItems;
    }
    save(path) {
        const absPath = path_1.resolve(path);
        fs_1.writeFileSync(absPath, this.toText(), "utf-8");
    }
    toText() {
        const message = this.secretKey.valueOf();
        return new pemEntry_1.PemEntry(this.label, message).toText();
    }
}
exports.ValidatorPEM = ValidatorPEM;
